shared/helpers/puppeteer.js

```diff
@@ -91,19 +90,42 @@ async function renderAndInitialise (page, componentName, options) {
     slot.innerHTML = htmlForSlot
   }, html)
 
+  // Call `beforeInitialisation` in a separate `$eval` call
+  // as running it inside the body of the next `evaluate`
+  // didn't provide a reliable execution
+  if (options.beforeInitialisation) {
+    await page.$eval('[data-module]', options.beforeInitialisation)
+  }
+
   // Run a script to init the JavaScript component
-  await page.evaluate(async (exportName, options) => {
+  //
+  // Use `evaluate` to ensure we run `document.querySelector` inside the
+  // browser, like users would, rather than rely on Puppeteer looking for the
+  // element which would cause an error in Jest-land rather than within the
+  // browser if the element is missingß
+  //
+  // Puppeteer returns very little information on errors thrown during
+  // `evaluate`, only a `name` that maps to the error class (and not its `name`
+  // property, which means we get a mangled value). As a workaround, we can
+  // gather and `return` the values we need from inside the browser, and throw
+  // them when back in Jest (to keep them triggering a Promise rejection)
+  const error = await page.evaluate(async (exportName, options) => {
     const $module = document.querySelector('[data-module]')
 
-    if (options.initialiser) {
-      options.initialiser($module)
-    }
-
     const namespace = await import('govuk-frontend')
-    /* eslint-disable-next-line no-new */
-    new namespace[exportName]($module, options.config)
+
+    try {
+      /* eslint-disable-next-line no-new */
+      new namespace[exportName]($module, options.config)
+    } catch ({ name, message }) {
+      return { name, message }
```

Comment:
So even if we re-throw (not return) on this line it doesn't work? 😣

Reply:
Indeed, the information was still lost when re-throwing 😣 I didn't think of checking what lands in the `pageerror` event, though so I'll give that a quick go if it can avoid us that juggling.

Reply:
No more luck looking at the `pageerror` and `error` events on the page. If I remove the `try/catch` in the code that's run by `page.evaluate`, no event gets triggered. The promise for `page.evaluate` simply rejects, with only a `name` property to the mangled class name of the error (see the following Jest output) 😞 

```diff
- Object {
-   "message": "GOV.UK Frontend is not supported in this browser",
-   "name": "GOVUKFrontendNotSupportedError",
+ Error {
+   "name": "n",
}
```

Reply:
What's weird is that puppeteer on its own seems to be happy with errors that have a custom name thrown from `page.evaluate`.

Taking the following script:

```mjs
import puppeteer from 'puppeteer';

// Launch the browser and open a new blank page
const browser = await puppeteer.launch({
  headless: 'new'
});
const page = await browser.newPage();

await page.goto('about:blank')

try {
  await page.evaluate(() => {
    class CustomError extends Error {
      name = 'CustomError'
    }

    throw new CustomError('Custom message in custom error')
  })
} catch (error) {
  console.log('Error class:', error.constructor.name)
  const {name, message, cause } = error
  console.log('Contents:', {name, message, cause})
  console.log(error)
}
```

The logs end up as expected (the class on the Node side can only be `Error` as the `CustomError` only exists inside the browser's page).

```sh
Error class: Error
Contents: {
  name: 'CustomError',
  message: 'Custom message in custom error',
  cause: undefined
}
Error [CustomError]: Custom message in custom error
```

So there's something in between that messes with our `page.evaluate` and requires us to add that workaround with `throw`.

I had a quick look in the sources of argos-ci/jest-puppeteer (especially jest-environment-puppeteer), but didn't notice anything that would wrap `page.evaluate` to mess with the error thrown. The page seems to be a regular Puppeteer page 😔

Reply:
Is the name-shortening to `n` [from Terser](https://github.com/alphagov/govuk-frontend/pull/4030#discussion_r1284234033)?

Reply:
Yup, that's what ends up in the `govuk-frontend.min.js` that's being built in `dist` and then loaded by the review app 😊